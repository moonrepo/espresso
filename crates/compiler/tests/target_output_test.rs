mod utils;

use espresso_common::EsTarget;
use espresso_compiler::Compiler;
use espresso_package::Package;
use starbase_sandbox::{assert_snapshot, create_sandbox};
use utils::read_file;

macro_rules! test_target {
    ($method:ident, $target:expr) => {
        #[tokio::test]
        async fn $method() {
            let sandbox = create_sandbox("syntax");
            let package = Package::new(sandbox.path()).unwrap();
            let compiler = Compiler::new(&package).unwrap();
            let out_dir = compiler.compile($target).await.unwrap();

            assert_snapshot!(read_file(out_dir.join("index.mjs")));
        }
    };
}

mod target_output {
    use super::*;

    test_target!(es2015, EsTarget::Es2015);
    test_target!(es2016, EsTarget::Es2016);
    test_target!(es2017, EsTarget::Es2017);
    test_target!(es2018, EsTarget::Es2018);
    test_target!(es2019, EsTarget::Es2019);
    test_target!(es2020, EsTarget::Es2020);
    test_target!(es2021, EsTarget::Es2021);
    test_target!(es2022, EsTarget::Es2022);

    #[tokio::test]
    async fn supports_legacy_decorators() {
        let sandbox = create_sandbox("syntax-legacy-decorators");
        let package = Package::new(sandbox.path()).unwrap();
        let compiler = Compiler::new(&package).unwrap();
        let out_dir = compiler.compile(EsTarget::Es2018).await.unwrap();

        assert_snapshot!(read_file(out_dir.join("index.mjs")));
        assert_snapshot!(read_file(out_dir.join("other.mjs")));
    }
}
