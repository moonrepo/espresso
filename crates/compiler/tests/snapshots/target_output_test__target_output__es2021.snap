---
source: crates/compiler/tests/target_output_test.rs
expression: "read_file(out_dir.join(\"index.mjs\"))"
---
function _check_private_redeclaration(obj, privateCollection) {
    if (privateCollection.has(obj)) {
        throw new TypeError("Cannot initialize the same private elements twice on an object");
    }
}
function _class_apply_descriptor_set(receiver, descriptor, value) {
    if (descriptor.set) {
        descriptor.set.call(receiver, value);
    } else {
        if (!descriptor.writable) {
            throw new TypeError("attempted to set read only private field");
        }
        descriptor.value = value;
    }
}
function _class_extract_field_descriptor(receiver, privateMap, action) {
    if (!privateMap.has(receiver)) {
        throw new TypeError("attempted to " + action + " private field on non-instance");
    }
    return privateMap.get(receiver);
}
function _class_private_field_init(obj, privateMap, value) {
    _check_private_redeclaration(obj, privateMap);
    privateMap.set(obj, value);
}
function _class_private_field_set(receiver, privateMap, value) {
    var descriptor = _class_extract_field_descriptor(receiver, privateMap, "set");
    _class_apply_descriptor_set(receiver, descriptor, value);
    return value;
}
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
var _private = new WeakMap();
export class ClassName {
    method() {
        _class_private_field_set(this, _private, false);
    }
    constructor(){
        _define_property(this, "field", 123);
        _class_private_field_init(this, _private, {
            writable: true,
            value: true
        });
        _define_property(this, "anonMethod", ()=>{
            this.field = 456_789;
        });
    }
}
_define_property(ClassName, "field", 'abc');
(()=>{
    ClassName.field = 'xyz';
})();
export async function asyncFunc(...args) {
    return Promise.resolve();
}
export function* genFunc(value = 'abc') {
    yield 1;
}
export async function* asyncGenFunc() {
    yield 'abc';
}
const obj = {
    a: 1,
    b: 2,
    c: 3
};
const { a, b, c } = obj;
const [d, e, f] = [
    1,
    2,
    3
];
const nullable = 1 ?? 2 ?? 3;
const optional = obj?.a || 0;
let assign = 0;
assign ||= a;
assign &&= b;
assign ??= b;
try {
    await import('path');
} catch  {}
// Hrmmm
export function Component() {
    return /*#__PURE__*/ React.createElement("div", null);
} /*#__PURE__*/ 

